lib/sift/filter.rb

```diff
@@ -73,12 +70,44 @@ def valid_scope_params?(scope_params)
       scope_params.is_a?(Array) && scope_params.all? { |symbol| symbol.is_a?(Symbol) }
     end
 
-    def handler
-      parameter.handler
+    def handler(value, params)
+      if should_apply_default?(value)
+        default
+      elsif type == :scope
+        Sift::ScopeHandler.new(value, mapped_scope_params(params), parameter, scope_types)
+      else
+        Sift::WhereHandler.new(value, parameter)
+      end
+    end
+
+    def validate_scope_types!
+      return if scope_types.empty?
+
+      unless Sift::TypeValidator.new(param, scope_types.first).valid_type?
+        raise ArgumentError, "scope_types must contain a valid filter type for the scope parameter"
+      end
+      return if scope_types.size == 1
+
+      if scope_types.size > 2 || !valid_scope_option_types!(scope_types[1])
+        raise ArgumentError, "type: scope: expected to have this structure: [type, {#{scope_params.map { |sp| "#{sp}: type" }.join(', ')}}]"
+      end
+    end
+
+    def valid_scope_option_types!(hash)
+      valid_form = hash.respond_to?(:keys) && hash.all? { |key, value| Sift::TypeValidator.new(key, value).valid_type? }
+      if valid_form && scope_params.empty?
+        # default scope_params
+        @scope_params = hash.keys
+      end
+      valid_form && (hash.keys - scope_params).empty?
     end
 
-    def supports_ranges?
-      parameter.supports_ranges?
+    def validation_type
+      if type != :scope || scope_types.empty?
+        type
+      else
+        scope_types.first
```

Comment:
Only a single scope type is allowed right? If thats the case should the argument be changed to `scope_type` with a default value of nil?

Reply:
Another thought is that scope_type param is unrelated to other filters. Should we create a different filter class to have better argument cohesion for scopes and so that way we aren't complicating other filters with logic specific to scopes?

Reply:
scope_types can have something like [:int, {foo: :datetime}] or [:int]. I can definitely see making some kind of inheritance / mixin relationship between WhereFilter and ScopeFilter. I think they both support the `default:` option so there's some overlap. I totally agree that the frequently unused `scope_params` and `scope_types` arguments are a smell.

Reply:
What do you think of this @amayer171 ?
https://github.com/procore/sift/pull/49

Reply:
merged...